cellarium/ml/models/cellarium_gpt.py

```diff
@@ -0,0 +1,774 @@
+        attn_mult:
+            Multiplier for the attention scores.
+        input_mult:
+            Multiplier for the input embeddings.
+        backend:
+            Backend for the attention computation.
+    """
+
+    def __init__(
+        self,
+        n_vars: int,
+        d_model: int = 128,
+        d_ffn: int = 256,
+        n_heads: int = 4,
+        n_blocks: int = 6,
+        dropout: float = 0.02,
+        use_bias: bool = False,
+        attn_mult: float = 1.0,
+        input_mult: float = 1.0,
+        backend: Literal["keops", "torch", "math", "flash", "mem_efficient"] = "keops",
+    ) -> None:
+        super().__init__()
+        # +1 token for total_mrna_umis
+        self.id_embedding = nn.Embedding(n_vars + 1, d_model)
+        # continuous value embedding
+        self.value_embedding = ValueEmbedding(d_model, use_bias=use_bias)
+        # mask token for target values
+        self.mask_embedding = nn.Embedding(1, d_model)
+
+        self.n_blocks = n_blocks
+        self.blocks = nn.ModuleList(
+            [TransformerBlock(d_model, d_ffn, n_heads, dropout, attn_mult, use_bias, backend) for _ in range(n_blocks)]
+        )
+
+        self.input_mult = input_mult
+
+    def forward(
+        self,
+        ids_nc: torch.Tensor,
+        values_nc: torch.Tensor,
+        labels_mask_nc: torch.Tensor,
+        prefix_len_n: torch.Tensor,
+        attention_type: Literal["block", "block_diagonal", "full"],
+    ) -> torch.Tensor:
+        device = ids_nc.device
+
+        id_tokens_ncd = self.id_embedding(ids_nc)
+        value_tokens_ncd = self.value_embedding(torch.log1p(values_nc.float()).unsqueeze(-1))
+        mask_token = self.mask_embedding(torch.tensor(0, device=device))
+        value_tokens_ncd[labels_mask_nc] = mask_token
+
+        hidden_state_ncd = (id_tokens_ncd + value_tokens_ncd) * self.input_mult
+        for block in self.blocks:
+            hidden_state_ncd = block(
+                hidden_state_ncd,
+                prefix_len_n=prefix_len_n,
+                attention_type=attention_type,
+            )
+
+        return hidden_state_ncd
+
+
+class NegativeBinomialHead(nn.Module):
+    """
+    Negative binomial head.
+
+    Args:
+        d_model:
+            Dimensionality of the embeddings and hidden states.
+        use_bias:
+            Whether to use bias in the linear transformations.
+        output_mult:
+            Multiplier for the output embeddings.
+    """
+
+    def __init__(self, d_model: int, use_bias: bool, output_mult: float) -> None:
+        super().__init__()
+        self.dense = nn.Linear(d_model, 2, bias=use_bias)
+        self.output_mult = output_mult
+
+    def forward(self, hidden_state_ncd) -> tuple[torch.Tensor, torch.Tensor]:
+        output = self.dense(hidden_state_ncd) * self.output_mult
+        mu_nc = output[..., 0].exp()
+        theta_nc = output[..., 1].exp()
+        return mu_nc, theta_nc
+
+
+class CellariumGPT(CellariumModel):
```

Comment:
Can you make it a `PredictMixin` and implement `predict()` ?

Reply:
Although I'm thinking that, analogous to Geneformer, `predict()` would return the gene embedding vectors.  But I'm not sure that exactly makes sense here... you might also imagine that "predict" would return the negative binomial distributions.

My interest was in having a common sort of interface I could use to extract gene embeddings, whether or not the model was a Geneformer model or a CellariumGPT model.  I was previously using `.predict()` to do this from Geneformer, so I thought it would be nice if CellariumGPT worked the same way.

Reply:
See #152 if you're interested in this idea

Reply:
I've been locally implementing different versions of `predict` method depending on what I wanted to analyze. Since I was changing it a lot decided not to add it here. But it should be something similar to Geneformer predict that returns a dictionary and we can have boolean flags that can be used to control what it returns.